core/src/main/scala/kafka/log/UnifiedLog.scala

```diff
@@ -1336,34 +1336,39 @@ class UnifiedLog(@volatile var logStartOffset: Long,
         val maxTimestampSoFar = latestTimestampSegment.readMaxTimestampAndOffsetSoFar
         // lookup the position of batch to avoid extra I/O
         val position = latestTimestampSegment.offsetIndex.lookup(maxTimestampSoFar.offset)
-        latestTimestampSegment.log.batchesFrom(position.position).asScala
+        val timestampAndOffsetOpt = latestTimestampSegment.log.batchesFrom(position.position).asScala
           .find(_.maxTimestamp() == maxTimestampSoFar.timestamp)
           .flatMap(batch => batch.offsetOfMaxTimestamp().asScala.map(new TimestampAndOffset(batch.maxTimestamp(), _,
             Optional.of[Integer](batch.partitionLeaderEpoch()).filter(_ >= 0))))
+        OffsetResultHolder(timestampAndOffsetOpt)
```

Comment:
Went over [KIP-734](https://cwiki.apache.org/confluence/display/KAFKA/KIP-734%3A+Improve+AdminClient.listOffsets+to+return+timestamp+and+offset+for+the+record+with+the+largest+timestamp), the purpose of MAX_TIMESTAMP is to get the offset of the record with highest timestamp in the partition:

```
Used to retrieve the offset with the largest timestamp of a partition as message timestamps can be specified client side this may not match the log end offset returned by LatestSpec
```

With remote storage enabled, all the passive segments might be uploaded to remote and the local-log might contain only one **empty** active segment. We have to handle the `MAX_TIMESTAMP` case for remote storage. Thanks for catching this! Filed KAFKA-17552 to track this issue separately.

Reply:
I went over #15621 to see how do we handle the MAX_TIMESTAMP case for normal topics.

Now, we maintain the [shallowOffsetOfMaxTimestampSoFar](https://sourcegraph.com/github.com/apache/kafka@trunk/-/blob/storage/src/main/java/org/apache/kafka/storage/internals/log/LogSegment.java?L212) in LogSegment instead of the real-max-timestamp-offset. While uploading the LogSegment to remote, we create the [RemoteLogSegmentMetadata](https://sourcegraph.com/github.com/apache/kafka@trunk/-/blob/storage/api/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogSegmentMetadata.java) event which holds the metadata information about the segment. 

Even, if we pass the `shallowOffsetOfMaxTimestampSoFar` in the RemoteLogSegmentMetadata event, we have to download the remote-log-segment to find the real `offsetOfMaxTimestampSoFar`. This will increase the load on remote storage assuming that the original intention of the KIP-734 is to `Confirming topic/partition "liveness"` which means the Admin client will repeatedly invoke the list-offsets API for MAX_TIMESTAMP. 

The predominant case to find the "Confirming topic/partition livness" is to query the local-log which will work as expected. For MAX_TIMESTAMP, when enabled with remote storage, the results can go wrong when:

1. Assuming the timestamps are monotonic and all the passive segments are uploaded to remote and deleted from local. The only active segment in local disk is empty (post the log roll), then the max-timestamp will returned as "-1".
2. Assuming the timestamps are non-monotonic, then the (timestamp, offset) returned by the API may not be "TRUE" (max-timestamp, max-timestamp-offset) as it considers the segments only in the local-log.

Should we handle/drop the MAX_TIMESTAMP case for topics enabled with remote storage? This can cause high load:

1. On the RemoteLogMetadataManager, as we have to scan all the uploaded segment events to find the max-timestamp. And, the search should always proceed from remote to local storage.
2. On the RemoteStorageManager, when there exists the MAX_TIMESTAMP record in the remote storage, then we have to download that segment (few bytes) repeatedly.

Reply:
> Even, if we pass the shallowOffsetOfMaxTimestampSoFar in the RemoteLogSegmentMetadata event, we have to download the remote-log-segment to find the real offsetOfMaxTimestampSoFar.

IMHO, we don't need to pass `shallowOffsetOfMaxTimestampSoFar` to `RemoteLogSegmentMetadata` as `RemoteLogSegmentMetadata` has `maxTimestampMs` already, so the basic behavior is listed below (same as you described I'd say)
1. find the max timestamp from local segments
2. query `findOffsetByTimestamp` by max timestamp from step_1
3. compare the timestamp of record from remote to local to pick up correct offset

>  This will increase the load on remote storage assuming that the original intention of the KIP-734 is to Confirming topic/partition "liveness" which means the Admin client will repeatedly invoke the list-offsets API for MAX_TIMESTAMP.

The impl of KIP-734 was wrong because we don't loop all records in all path (because of cost issue). Hence,  we rename the `offsetOfMaxTimestampSoFar` to `shallowOffsetOfMaxTimestampSoFar` based on true story.

> Should we handle/drop the MAX_TIMESTAMP case for topics enabled with remote storage? This can cause high load:

that is a acceptable approach. We can REJECT the MAX_TIMESTAMP request for now as it is rare operation. Or we can make the call an async op too as it needs to iterate all metadata of remote segments.